Add RST spellcheck

oraNod opened the pull request:
Uses https://github.com/codespell-project/codespell to look for common spelling mistakes.

oraNod commented:
I just noticed that `pyproject.toml` is ignored so this PR doesn't include the codespell config.

```
[tool.codespell]
skip = '_build,_static,./venv,Makefile,*.sphinx,*.py,*.dot,*.inv,*.js'
count = ''
check-filenames = true
quiet-level = 3
ignore-words-list = 'aci'
```

@samccann Should we really ignore this file?

gotmax23 commented:
@oraNod, we can't have a pyproject.toml here, as it's overwritten by ansible-core's. Can codespell store its config in a different file? If needed, we can adjust clone-core.py to merge our own changes into core's pyproject.toml when syncing, but that of course adds extra complexity.

oraNod commented:
> @oraNod, we can't have a pyproject.toml here, as it's overwritten by ansible-core's. Can codespell store its config in a different file? If needed, we can adjust clone-core.py to merge our own changes into core's pyproject.toml when syncing, but that of course adds extra complexity.

Thanks for the reminder @gotmax23 I forgot about the core pyproject.toml. I've updated to use a codespell.toml file because I kind of like that convention. Maybe we can consolidate this with some other configuration by making it more generic. Open to any suggestions. Thanks.

oraNod commented:
Hmm. I guess we should add a requirements file under `tests` and maybe some docs so we can tell users how to fix spelling checks whenever ci fails against a pull request.

It would be `nox -s spelling -- -w`

I'll work on that tomorrow though.

gotmax23 commented on .github/workflows/ci.yaml:
```diff
@@ -59,3 +59,6 @@ jobs:
   nox:
     uses: ./.github/workflows/reusable-nox.yml
+
+  spell-check:
```
You can add this to the matrix in reusable-nox.yml instad of creating a whole new file:

https://github.com/ansible/ansible-documentation/blob/18a7692cff6ac5f8e8bd65fc1e70a8504d85892e/.github/workflows/reusable-nox.yml#L12-L19

oraNod replied:
good idea. I wasn't really sure if should go in there tbh.

gotmax23 commented on noxfile.py:
```diff
@@ -83,3 +83,19 @@ def pip_compile(session: nox.Session, req: str):
+
+
+@nox.session
+def spelling(session: nox.Session):
```
You should also add `session.notify("spelling")` to the lint session.

gotmax23 commented on .codespell.toml:
```diff
@@ -0,0 +1,5 @@
+[codespell]
+skip = '_build,_static,./venv,Makefile,*.sphinx,*.py,*.dot,*.inv,*.js'
```
We should ignore all .po files and porting_guides as well.

gotmax23 commented on .codespell.toml:
```diff
@@ -0,0 +1,5 @@
+[codespell]
+skip = '_build,_static,./venv,Makefile,*.sphinx,*.py,*.dot,*.inv,*.js'
```
Can these options be specified using normal TOML lists or do they have to use comma separated strings?

oraNod replied:
no, TOML list doesn't seem to work so I've changed to use `.codespellrc` instead

felixfontein commented:
I guess the auto-generated porting guides should also be skipped, at least the currently active versions, since we have no control over their content. (Once they are no longer updated and we allow manual changes we can remove them from the ignore list and start fixing the errors in them :) )

oraNod commented:
Thanks for the reviews and comments @gotmax23 and @felixfontein Hopefully this is ready to go now.

Would it be better to fix spelling errors as part of this PR. Or should we just ignore the ci check and do a bulk update in a follow up so unrelated commits don't get squashed together?

gotmax23 commented on README.md:
```diff
@@ -7,3 +7,34 @@ This repository holds the ReStructuredText (RST) source, and other files, for us
+
+Run ``nox -s spelling -- -w`` to correct spelling errors.
+
+If you want to run spelling checks without ``nox`` do the following:
+
+    # Install codespell
+    python3 -m pip install -r tests/spelling.in -c tests/spelling.txt
+
+    # Check spelling
+    codespell docs/docsite
```
I would remove this part. I don't think we need to document two ways to do the same thing.

oraNod replied:
Sounds good. I'll add a link to the codespell project page so users can figure out how to do their own thing if they wish.

gotmax23 commented on tests/spelling.in:
```diff
@@ -0,0 +1 @@
+codespell
```
Can you fix the no newline at EOF error?

gotmax23 commented on tests/spelling.txt:
You need to edit

https://github.com/ansible/ansible-documentation/blob/ad2deae89b03b29ab236a6295cc9537a6e715f1e/.github/workflows/pip-compile-dev.yml#L38-L41

and add `'pip-compile-3.10(spelling)'` so this gets updated as part of the weekly dep bump PRs.

gotmax23 commented on README.md:
```diff
@@ -7,3 +7,34 @@ This repository holds the ReStructuredText (RST) source, and other files, for us
+
+## Using nox
+
+This project includes a ``nox`` configuration to automate checks and other functions.
```
Markdown only uses one set of backticks unlike RST:

```suggestion
This project includes a `nox` configuration to automate checks and other functions.
```

gotmax23 commented on README.md:
```diff
@@ -7,3 +7,34 @@ This repository holds the ReStructuredText (RST) source, and other files, for us
+## Using nox
+
+This project includes a ``nox`` configuration to automate checks and other functions.
+We recommend that you use ``nox`` virtual sessions to run checks locally before you submit a pull request.
```
```suggestion
You should use `nox` to run checks locally before you submit a pull request.
```

gotmax23 commented on README.md:
```diff
@@ -7,3 +7,34 @@ This repository holds the ReStructuredText (RST) source, and other files, for us
+This project includes a ``nox`` configuration to automate checks and other functions.
+We recommend that you use ``nox`` virtual sessions to run checks locally before you submit a pull request.
+
+Install ``nox``.
+
+    # Install nox with pip
+    python3 -m pip install nox
+
+    # Install nox on MacOS
+    brew install nox
```
``` suggestion
Install `nox` using `python3 -m pip install nox` or your distribution's package manager.
```

gotmax23 reviewed (commented):
Thanks @oraNod! I have some final comments.

gotmax23 commented:
> Would it be better to fix spelling errors as part of this PR. Or should we just ignore the ci check and do a bulk update in a follow up so unrelated commits don't get squashed together?

I would fix it in this PR. If you want to push changes to fix the errors, I can handle squashing the PR into two commits, one to add the config and another to fix the errors.

gotmax23 commented:
> Ansible Docsite CI / nox / Run nox formatters_check session (pull_request) Failing after 19s 

You need to run `nox -e formatters` and push the results to fix some black errors.

oraNod commented:
> > Ansible Docsite CI / nox / Run nox formatters_check session (pull_request) Failing after 19s
> 
> You need to run `nox -e formatters` and push the results to fix some black errors.

Hmm. I thought I did this but just need to make it a habit before every time I push. Thanks for pointing it out though.

gotmax23 commented:
There remains some CI failures for you to fix :).

oraNod committed:
Add docs spelling check

* Add codespell config to noxfile and ci
* Update README with details on spelling checks

Co-authored-by: Maxwell G <9920591+gotmax23@users.noreply.github.com>

oraNod committed:
fix spelling issues

oraNod commented:
> There remains some CI failures for you to fix :).

They are the gifts that keep on giving. Should be fine now.

samccann commented:
I tried this branch locally and had a couple of comments:
1 - if `nox` is required now, should it be added to the requirements files as well?
2 - I forced a spelling error in a couple of files and did 'nox -s spelling' and it didn't pick up any errors. Not sure if I'm forgetting a step here?

```
(devel) samccann:ansible-documentation$ nox -s spelling
nox > Running session spelling
nox > Creating virtual environment (virtualenv) using python3 in .nox/spelling
nox > python -m pip install -r tests/spelling.in
nox > codespell docs/docsite
Used config files:
    1: .codespellrc
0
nox > Session spelling was successful.
```

oraNod commented:
@samccann Thanks for taking a look. `nox` isn't a requirement like the packages we need to build the docsite. it's kind of like `pip` or `wheels`. it's tooling you need to install before you install requirements. but it isn't strictly necessary. you can install `codespell` directly and run the spelling check without `nox`.

using `nox` means running the spelling check - and other sessions - are consistent across distros etc. I have added a section to the README about installing `nox` and can update that for other sessions (but in a follow up)

gotmax23 commented:
>  1 - if nox is required now, should it be added to the requirements files as well?

No, the requirements files are used *by nox* to install dependencies for each test session, but users should simply run `pip install nox` to get nox itself.

oraNod commented:
also @samccann it can depend where you add the spelling errors because `codespell` skips some directories, including `collections` and `style_guide`

it can also depend on the dictionary. `codespell` might not catch every spelling error but we can change dictionaries or contribute spelling error suggestions to the default one if needed.

what words did you test with?

webknjaz commented:
> > 1 - if nox is required now, should it be added to the requirements files as well?
> 
> No, the requirements files are used _by nox_ to install dependencies for each test session, but users should simply run `pip install nox` to get nox itself.

Technically, having a *separate* lockfile just for the external tooling (like nox itself) is not such a wild idea. It's not a widely used approach, but is nice to have in general. Definitely out of the scope of this PR, though…

samccann commented:
@oraNod - I introduced spelling errors in multiple files and still not getting anything locally when I run nox -s spelling. I don't want to hold up the PR but can you try it locally? Maybe I have something messed up here. I added junk into the getting started guide and no errors show up.

oraNod commented:
> @oraNod - I introduced spelling errors in multiple files and still not getting anything locally when I run nox -s spelling. I don't want to hold up the PR but can you try it locally? Maybe I have something messed up here. I added junk into the getting started guide and no errors show up.

Please bear in mind that codespell won't catch _every_ spelling mistake and is only as good as the dictionary in use. Over time it would be good to contribute spelling errors to the codespell dictionary ourselves.

Here's an example of a spelling error that it caught: "direcory" Try putting this in the getting started guide then run again. You should see:

```
$ nox -s spelling
nox > Running session spelling
nox > Creating virtual environment (virtualenv) using python3 in .nox/spelling
nox > python -m pip install -r tests/spelling.in
nox > codespell docs/docsite
Used config files:
    1: .codespellrc
docs/docsite/rst/getting_started/get_started_ansible.rst:21: direcory ==> directory
1
nox > Command codespell docs/docsite failed with exit code 65
nox > Session spelling failed.
```

Now try this one: "Continu getting started" where the "e" is truncated. The spelling check doesn't fail here because "continu" isn't in one of the codespell dictionaries: https://github.com/codespell-project/codespell/tree/master/codespell_lib/data

We can either submit a pr to add that to the codespell dictionary or maybe create our own custom dictionary.

oraNod commented:
@samccann I'll also add some more dictionaries to the codespell config while we're here. One that I haven't added yet, because I wanted to do it separately as a follow up is https://github.com/codespell-project/codespell/blob/master/codespell_lib/data/dictionary_usage.txt

This will catch problematic terms like "whitelist". I guess we could go ahead and fix that now if you think.

oraNod commented:
Gave it a final force push to revert some spelling fixes that should have been ignored. This should be ready to go if anyone wants to approve.

samccann commented:
Tested locally and seems to work. @oraNod - I don't know if you want anyone else to look at this?  Also, my recommendation for now is that it stays in devel for a couple of weeks so we know it's not causing problems before we backport to stable-2.16?  Or we could keep it just on devel as that's where all PRs start out for the most part?

oraNod commented:
@samccann I think keeping it on `devel` sounds good. Makes sense to me. I wouldn't mind a second approval from either @gotmax23 or @felixfontein to make sure we have someone from the steering committee give "the blessing".

gotmax23 commented:
I'll give it another review.

> my recommendation for now is that it stays in devel for a couple of weeks so we know it's not causing problems before we backport to stable-2.16? Or we could keep it just on devel as that's where all PRs start out for the most part?

I like to keep the CI configuration synced across branches as much as possible, but I'm fine with "incubating" this on devel for a bit.

gotmax23 commented on README.md:
```diff
@@ -7,3 +7,22 @@ This repository holds the ReStructuredText (RST) source, and other files, for us
+
+Install `nox` using `python3 -m pip install nox` or your distribution's package manager.
+
+Run `nox --list` from the repository root to view available sessions.
```
```suggestion
Run `nox --list` from the repository root to view available sessions.

Run `nox` with no arguments to execute the default sessions.
```

gotmax23 commented on .codespellrc:
```diff
@@ -0,0 +1,8 @@
+[codespell]
+skip = *.po,_build,_static,porting_guide*,collections,style_guide,
```
I prefer the newline-separated INI list format. I think it's easier to read, but that's up to you :). At least, it'd be good if there were spaces between items in the comma-separated format.

```suggestion
skip = 
    *.po
    _static
    build
    collections
    porting_guide*
    style_guide
```

(I sorted the items in the process)

oraNod replied:
Thanks @gotmax23 one thing that tripped me up was putting a space between items meant they were ignored by codespell. Let's commit this and see if it breaks anything. Thanks for the sort too.

gotmax23 commented on .codespellrc:
```diff
@@ -0,0 +1,8 @@
+quiet-level = 3
+ignore-words = .codespellignorewords
+exclude-file = .codespellignorelines
+builtin = clear,rare,names,en-GB_to_en-US,informal,code
```
```suggestion
builtin = 
    clear
    code
    en-GB_to_en-US,
    informal
    names
    rare
```

(I sorted the items in the process)

oraNod replied:
@gotmax23 Unfortunately it looks like codespell only respects the comma separated list (with no space characters).

gotmax23 replied:
Yeah, the config parsing code is very naive. Let's leave it as is.

/resolve

gotmax23 reviewed (approved):
I have some INI quibbles and another README suggestion, but otherwise this looks good to me.
Thanks for working on it!
Let's merge this using the newly-enabled merge commit option :).

gotmax23 commented:
You'll have to resolve conflicts now that https://github.com/ansible/ansible-documentation/commit/b9323fbe54986d4a8ea941653bae1bb92597b4c8 has been merged 😀.

oraNod commented:
> You'll have to resolve conflicts now that [b9323fb](https://github.com/ansible/ansible-documentation/commit/b9323fbe54986d4a8ea941653bae1bb92597b4c8) has been merged 😀.

Yes, I'm aware. Thanks.
